### PR TITLE
Obtain search results for a particular subsite when the request comes from a query

### PR DESCRIPTION
**Pull request by silverstripe-elliot**
When attempting to do this in project code, SearchVariantSubsites kicks in and replaces any filter you've applied with the current subsite ID, which prevents you from searching for filtered results in another subsite.

This fix prevents the module from doing this if a filter is already applied on the query

### CODE REVIEW

**Comment by silverstripe-elliot** [comment]
Any updates on this?

**Comment by dhensby** [comment]
This looks good. Do you think you could add a test?

**Comment by silverstripe-elliot** [comment]
Hey @dhensby! There wasn't an existing test for SearchVariantSubsites, so I've created a new one. Since there aren't a lot of examples to look at, I've limited the scope of the test to only cover my changes. I hope this is OK!

**Comment by dhensby** [comment]
That's perfect, thanks.

It looks like [the test failure is related](https://travis-ci.org/silverstripe/silverstripe-fulltextsearch/jobs/216211413) and I've left some comments.

File: tests/SearchVariantSubsitesTest.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private static $index = null;
+
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        // Check versioned available
+        if (!class_exists('Subsite')) {
+            return $this->markTestSkipped('The versioned decorator is not installed');
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: Probably should say that the subsite module is not installed?

File: tests/SearchVariantSubsitesTest.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        //typical behaviour: nobody is explicitly filtering on subsite, so the search variant adds a filter to the query
+        $this->assertTrue(empty($query->require['_subsite']));
+        $variant = new SearchVariantSubsites();
+        $variant->alterDefinition('SearchUpdaterTest_Container', $index);
+        $variant->alterQuery($query, $index);
+
+        //check that the "default" query has been put in place: it's not empty, and we're searching on Subsite ID:0 and
+        // an object of SearchQuery::missing
+        $this->assertFalse(empty($query->require['_subsite']));
+        $this->assertTrue($query->require['_subsite'][0] == 0);
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertEquals(0, $query->require['_subsite'][0])`

File: tests/SearchVariantSubsitesTest.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        $this->assertTrue($query->require['_subsite'][1] instanceof SearchQuery::$missing);
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertInstanceOf(SearchQuery::$missing, $query->require['_subsite'][1])`

File: tests/SearchVariantSubsitesTest.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+
+    public function testQueryIsAlteredWhenSubsiteIsSet()
+    {
+        //now we want to test if somebody has already applied the _subsite filter to the query
+        $index = new SolrIndexTest_FakeIndex();
+        $query = new SearchQuery();
+
+        //check that _subsite is not applied yet
+        $this->assertTrue(empty($query->require['_subsite']));
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertEmpty($query->require['_subsite'])` Although, if the key may not exist then maybe you'll have to leave it as is - please add a comment if that's the case.

File: tests/SearchVariantSubsitesTest.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        //apply the subsite filter on the query (for example, if it's passed into a controller and set before searching)
+        //we've chosen an arbirary value of 2 here, to check if it is changed later
+        $query->filter('_subsite', 2);
+        $this->assertFalse(empty($query->require['_subsite']));
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertNotEmpty($query->require['_subsite'])`

File: tests/SearchVariantSubsitesTest.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        //apply the search variant's definition and query
+        $variant = new SearchVariantSubsites();
+        $variant->alterDefinition('SearchUpdaterTest_Container', $index);
+
+        //the protected function isFieldFiltered is implicitly tested here
+        $variant->alterQuery($query, $index);
+
+        //confirm that the query has been altered, but NOT with default values
+        //first check that _subsite filter is not empty
+        $this->assertFalse(empty($query->require['_subsite']));
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertNotEmpty($query->require['_subsite'])`

File: tests/SearchVariantSubsitesTest.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        //subsite filter first value is not 0
+        $this->assertFalse($query->require['_subsite'][0] == 0);
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertNotEquals(0, $query->require['_subsite'][0])`

File: tests/SearchVariantSubsitesTest.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        //subsite filter second value is not SearchQuery::missing
+        $this->assertFalse($query->require['_subsite'][1] instanceof SearchQuery::$missing);
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertNotInstanceOf(SearchQuery::$missing, $query->require['_subsite'][1])`

File: tests/SearchVariantSubsitesTest.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+        //subsite filter has been modified with our arbitrary test value. The second value is not set
+        //this proves that the query has not been altered by the variant
+        $this->assertTrue($query->require['_subsite'][0] == 2);
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertEquals(2, $query->require['_subsite'][0])`

File: tests/SearchVariantSubsitesTest.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        $this->assertTrue(empty($query->require['_subsite'][1]));
```
Comment (dhensby) [inline tests/SearchVariantSubsitesTest.php]: `$this->assertEmpty($query->require['_subsite'][1])`

**Comment by silverstripe-elliot** [comment]
Thanks @dhensby! All of these issues have been fixed, and I've addressed the reason for the test failure

**Pull request merged**